src/app/codegen-interaction-model/Model.h

```diff
@@ -0,0 +1,52 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/interaction-model/Model.h>
+
+namespace chip {
+namespace app {
+namespace CodegenDataModel {
+
+class Model : public chip::app::InteractionModel::Model
+{
+public:
+    /// Generic model implementations
+    CHIP_ERROR Shutdown() override { return CHIP_NO_ERROR; }
+
+    CHIP_ERROR ReadAttribute(const InteractionModel::ReadAttributeRequest & request, InteractionModel::ReadState & state,
+                             AttributeValueEncoder & encoder) override;
+    CHIP_ERROR WriteAttribute(const InteractionModel::WriteAttributeRequest & request, AttributeValueDecoder & decoder) override;
+    CHIP_ERROR Invoke(const InteractionModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                      InteractionModel::InvokeReply & reply) override;
+
+    /// attribute tree iteration
+    EndpointId FirstEndpoint() override;
+    EndpointId NextEndpoint(EndpointId before) override;
+
+    InteractionModel::ClusterEntry FirstCluster(EndpointId endpoint) override;
```

Comment:
Similar question. How do you see this used?

Reply:
This is looking to implement wildcard expansion that is done in https://github.com/project-chip/connectedhomeip/blob/master/src/app/AttributePathExpandIterator.cpp#L176.

That object contains a `current path` as well as highly-compact indexes to iterate through: https://github.com/project-chip/connectedhomeip/blob/master/src/app/AttributePathExpandIterator.h#L112

The compactness seems needed because the expand iterator exists in every ReadHandler and we have multiple ReadHandlers based on the number of supported reads per fabric and fabrics, so RAM usage gets multipled a lot. I will already have to add a pointer to the cluster data provider (called 'Model' here but will rename soon based on review feedbac) so I was afraid adding more iterator bits also because we do not currently really have `cancellable iterators` concepts. So I left it bare with the option to update it later. 

This is what pahole says about existing AttributePathExpandIterator sizes:

```
class AttributePathExpandIterator {
public:

    struct SingleLinkedListNode<chip::app::AttributePathParams> * mpAttributePath; /*     0     8 */
    struct ConcreteAttributePath mOutputPath;        /*     8    12 */
    uint16_t                   mEndpointIndex;       /*    20     2 */
    uint16_t                   mEndEndpointIndex;    /*    22     2 */
    uint16_t                   mAttributeIndex;      /*    24     2 */
    uint16_t                   mEndAttributeIndex;   /*    26     2 */
    uint8_t                    mClusterIndex;        /*    28     1 */
    uint8_t                    mEndClusterIndex;     /*    29     1 */
    uint8_t                    mGlobalAttributeIndex; /*    30     1 */
    uint8_t                    mGlobalAttributeEndIndex; /*    31     1 */

    /* size: 32, cachelines: 1, members: 10 */
    /* last cacheline: 32 bytes */
};
```

I think I can drop all indexes (looks like 12 bytes) however I have to add at least a model pointer (4 bytes).